feat: [M3-7185] - Add Drag and Drop support to the Rules Table

cpathipa opened the pull request:
## Description 📝
**Add drag/ drop functionality to Rules Table in Routes Tab**

## Major Changes 🔄
- Added drag and drop functionality to Rules Table.
- Implemented Move up / Move Down action behavior.
- Added useLoadBalancerRouteUpdateMutation.
- Unit tests for RulesTable component.

## Preview 📷
**Include a screenshot or screen recording of the change**
![image](https://github.com/linode/manager/assets/119517080/e4a1406f-2d8e-499c-9b63-6d116c4bc5f6)

## How to test 🧪
1. **How to setup test environment?**
- Check out this PR.
- `yarn dev`
- Ensure AGLB feature flag is **on** and MSW is **enabled**.
2. **How to verify changes?**
- Navigate to http://localhost:3000/loadbalancers/0/routes.
- Click on any route in the table.
- Test Drag and drop functionality in Rules Table.
- Confirm that Move up and Move down actions behaviors.
- Validate Put request integration for updateLoadbalancerRoute according to API spec. 

 
4. **How to run Unit or E2E tests?**
yarn test RulesTable

tyler-akamai commented:
Does it make sense to only be able to drag the objects hovering over the drag icon and not the entire row?

Currently, you can move the row by clicking anywhere on the row and dragging it.

tyler-akamai commented:
It also seems like the 'Move Up' and 'Move Down'
https://github.com/linode/manager/assets/139489745/eef38917-1407-48cf-8c23-c6c61ebeba84

tyler-akamai commented:
Verified:
 - [ ] Drag and drop functionality works as expected
 - [ ] 'Move Up' and 'Move Down' functionality works as expected
 - [ ] Validated Put request integration for updateLoadbalancerRoute according to API spec
 - [ ] yarn test RulesTable Passed

mjac0bs commented:
> Should it be so that the user can only drag/drop the row by hovering over _only_ the drag icon and not anywhere in row?
> 
> Currently, you can move the row by clicking anywhere on the row and dragging it, which might defeat the purpose of having the drag icon all the way to the left.

@tyler-akamai I've always thought of the drag icon as an indicator that the row is drag and droppable, not an area to activate that behavior. We'll want this table's behavior to be consistent with Firewall rules, which are drag-n-droppable by clicking anywhere in the row:

Firewall Rules, for reference

https://github.com/linode/manager/assets/114685994/e5f9c522-feca-4f5d-92e4-9a18f86797f0

tyler-akamai commented:
> > Should it be so that the user can only drag/drop the row by hovering over _only_ the drag icon and not anywhere in row?
> > Currently, you can move the row by clicking anywhere on the row and dragging it, which might defeat the purpose of having the drag icon all the way to the left.
> 
> @tyler-akamai I've always thought of the drag icon as an indicator that the row is drag and droppable, not an area to activate that behavior. We'll want this table's behavior to be consistent with Firewall rules, which are drag-n-droppable by clicking anywhere in the row:
> 
> Firewall Rules, for reference
> 
>  Screen.Recording.2023-10-06.at.7.25.02.AM.mov

Awesome, thank you for the clarification

mjac0bs reviewed (changes requested):
Nice work, this looks close!
- [X] Test Drag and drop functionality in Rules Table.
- [x] Confirm that Move up and Move down actions behaviors. -- I'm assuming that the weirdness with the `First` and `Last` label remaining unchanged, rather than following the rule when it is moved, is due to the mocks?
- [x] Validate Put request integration for updateLoadbalancerRoute according to API spec.
- [x] Tested drag and drop functionality with VoiceOver in Chrome for accessibility (see attached video) -- just want to note that we've had issues with the Firewall Rules drag and drop accessibility in the past with some browser/screen reader combinations, and we should be mindful of that here. I did not test other screen readers (e.g. JAWS) and browsers in this PR.

Changes requested:
- Fixing the Action button is slightly misaligned again in desktop view and gets very misaligned in mobile view when the table collapses.
- Using TextTooltip rather than a custom tooltip treatment.
- Fixing the alignment of the Service Target and Session Stickiness column. The other table columns have column headers that are left aligned. (See last screenshot.)

![Screenshot 2023-10-06 at 7 32 48 AM](https://github.com/linode/manager/assets/114685994/eb075c44-a61e-4352-8349-14bbe236eed7)

https://github.com/linode/manager/assets/114685994/f73308c8-b27f-4275-b9fe-748555adedc0

![Screenshot 2023-10-06 at 8 59 02 AM](https://github.com/linode/manager/assets/114685994/37a5a42b-1221-441a-9b88-53f8e78a3923)

bnussman-akamai commented on packages/manager/src/features/LoadBalancers/LoadBalancerDetail/RulesTable.tsx:
```diff
@@ -1,116 +1,308 @@
+                                }}
+                                aria-label={`Service Targets: ${rule.service_targets.length}`}
+                              >
+                                <Tooltip
+                                  title={
+                                    <>
+                                      {rule.service_targets.map(
+                                        ({ id, label }) => (
+                                          <div key={label}>
+                                            {label}:{id}
+                                          </div>
+                                        )
+                                      )}
+                                    </>
+                                  }
+                                >
+                                  <div
+                                    style={{
+                                      maxWidth: '30px',
+                                      textDecoration: 'underline',
+                                    }}
+                                  >
+                                    {rule.service_targets.length}
+                                  </div>
+                                </Tooltip>
```
I agree 100%

tyler-akamai commented on packages/manager/src/features/LoadBalancers/LoadBalancerDetail/RulesTable.tsx:
```diff
@@ -1,116 +1,308 @@
+      return;
+    }
+
+    if (result.destination) {
```
I believe this if statement is unnecessary. 

Just have:
```js
  const onDragEnd = (result: DropResult) => {
    if (
      !result.destination ||
      result.destination.index === result.source.index
    ) {
      return;
    }

    handleRulesReorder(result.source.index, result.destination!.index);
  };
```

cpathipa commented:
TY @mjac0bs Great feedback. Addressed all the feedback items.

- Fixing the Action button is slightly misaligned again in desktop view and gets very misaligned in mobile view when the table collapses.
- Using TextTooltip rather than a custom tooltip treatment.
- Fixing the alignment of the Service Target and Session Stickiness column. The other table columns have column headers that are left aligned. (screenshot.)

![image](https://github.com/linode/manager/assets/119517080/380d286b-758e-4c8c-aae7-13d796f6630b)

![image](https://github.com/linode/manager/assets/119517080/a0aae34d-d430-4ebb-a599-3ede4da798a2)

mjac0bs commented:
@cpathipa - The one thing I'm still seeing is the Session Stickiness and Service Target alignment inconsistencies:

![Screenshot 2023-10-06 at 11 38 42 AM](https://github.com/linode/manager/assets/114685994/cf94797e-1603-419a-903b-43a6e2504c5d)

cpathipa commented:
@mjac0bs forgot to adjust the table data columns. 
![image](https://github.com/linode/manager/assets/119517080/8bb6a25c-8611-4371-8b9c-4341f2e0c57c)

jdamore-linode reviewed (changes requested):
I'm getting test failures locally in `load-balancer-routes.spec.ts`. Hoping to circle back with an explanation or fix soon

mjac0bs commented on packages/manager/src/features/LoadBalancers/LoadBalancerDetail/RulesTable.tsx:
```diff
@@ -201,7 +201,7 @@ export const RulesTable = ({ loadbalancerId, route }: Props) => {
                                   ]
                                 }`}
-                                sx={{ ...sxItemSpacing, width: '20%' }}
+                                sx={{ ...sxItemSpacing, width: '15%' }}
```
The other columns look great! But this adjustment did have an unintended impact on the Action menu column: 

![Screenshot 2023-10-06 at 12 43 31 PM](https://github.com/linode/manager/assets/114685994/1b25e2bb-e2fc-414e-9bd0-dd86801fc05c)

cpathipa replied:
Yes, Actually column and body were using only 95% width instead of 100%. So adjusted that in 
a595192
![image](https://github.com/linode/manager/assets/119517080/ff02048a-3f55-40b2-b80d-7af5b1685130)

mjac0bs reviewed (approved):
Thank you for addressed feedback! 🚀 

This looks good, assuming that e2e issue gets resolved.

jdamore-linode commented:
@cpathipa I'm not sure why it's different, but I think the test is failing because of the way errors are being displayed in the Add Rule drawer.

`develop`:
<img width="637" alt="Screenshot 2023-10-06 at 3 54 20 PM" src="https://github.com/linode/manager/assets/97627410/7b1e2828-5d6d-478a-9f19-cb2433652092">

`M3-7185`:
<img width="596" alt="Screenshot 2023-10-06 at 4 12 17 PM" src="https://github.com/linode/manager/assets/97627410/592b1354-5354-44de-b19c-2423bf150141">

My best guess is the factory change is playing a role, but I'm admittedly playing catch up on how all this works.

jdamore-linode reviewed (approved):
Thanks @bnussman-akamai, nice work @cpathipa!

Confirmed the test is passing, and manually tested the rules drag and drop (which is really slick!)
